File: frontend/src/schemas/budgetPlanner.js
/**
 * Budget Planner V1.2 - API Schemas & Types
 * 
 * @module schemas/budgetPlanner
 * @description TypeScript/JSDoc type definitions for all budget planner data structures
 * Used for backend validation and frontend type safety
 */

// ============================================
// INPUT SCHEMAS
// ============================================

/**
 * @typedef {Object} IncomeInput
 * @property {number} monthly_income - Monthly income (₹)
 * @property {string} currency - Currency code (INR, USD, EUR, etc.)
 */

/**
 * @typedef {Object} CityInput
 * @property {string} [country] - Country name
 * @property {string} [state] - State/Province name
 * @property {string} [city] - City name
 * @property {string} city_tier - Tier (tier_1, tier_2, tier_3, other)
 * @property {number} col_multiplier - Cost-of-living multiplier
 */

/**
 * @typedef {Object} HouseholdInput
 * @property {number} family_size - Number of family members (1-10)
 * @property {string} lifestyle - Lifestyle type (minimal, moderate, comfort, premium)
 */

/**
 * @typedef {Object} ExpenseItem
 * @property {number} [rent] - Monthly rent (₹)
 * @property {number} [utilities] - Utilities (water, electricity, internet)
 * @property {number} [insurance] - Insurance premiums
 * @property {number} [medical] - Medical expenses
 * @property {number} [other] - Other fixed expenses
 */

/**
 * @typedef {Object} VariableExpenses
 * @property {number} [groceries] - Grocery expenses
 * @property {number} [transport] - Transport/fuel
 * @property {number} [subscriptions] - Subscriptions
 * @property {number} [entertainment] - Entertainment
 * @property {number} [shopping] - Shopping
 * @property {number} [dining_out] - Dining out
 * @property {number} [other] - Other variable
 */

/**
 * @typedef {Object} LoanInput
 * @property {number} principal - Original loan amount
 * @property {number} rate - Annual interest rate (%)
 * @property {number} tenure_months - Remaining tenure (months)
 * @property {string} [issuer] - Bank/NBFC name
 */

/**
 * @typedef {Object} SavingsGoal
 * @property {string} name - Goal name
 * @property {number} target - Target amount
 * @property {number} target_months - Timeline (months)
 * @property {number} priority - Priority (1-5)
 */

/**
 * @typedef {Object} BudgetGenerateRequest
 * @property {number} monthly_income - Monthly income
 * @property {string} currency - Currency code
 * @property {string} [country] - Country
 * @property {string} [state] - State
 * @property {string} [city] - City
 * @property {string} city_tier - Tier
 * @property {number} col_multiplier - COL multiplier
 * @property {number} family_size - Family size
 * @property {string} lifestyle - Lifestyle
 * @property {ExpenseItem} fixed_expenses - Fixed expenses
 * @property {VariableExpenses} variable_expenses - Variable expenses
 * @property {Array<LoanInput>} [loans] - Loan array
 * @property {Array<SavingsGoal>} [goals] - Savings goals
 * @property {string} mode - Budget mode (basic, aggressive_savings, smart_balanced)
 */

// ============================================
// OUTPUT SCHEMAS
// ============================================

/**
 * @typedef {Object} BudgetSplit
 * @property {number} needs_percent - Needs percentage
 * @property {number} wants_percent - Wants percentage
 * @property {number} savings_percent - Savings percentage
 */

/**
 * @typedef {Object} BudgetAmounts
 * @property {number} needs - Needs amount
 * @property {number} wants - Wants amount
 * @property {number} savings - Savings amount
 */

/**
 * @typedef {Object} NeedsCategory
 * @property {number} rent - Rent/Mortgage
 * @property {number} utilities - Utilities
 * @property {number} groceries - Groceries
 * @property {number} transport - Transport
 * @property {number} insurance - Insurance
 * @property {number} medical - Medical
 * @property {number} emi - EMI/Loan repayment
 * @property {number} other - Other needs
 */

/**
 * @typedef {Object} WantsCategory
 * @property {number} dining - Dining out
 * @property {number} entertainment - Entertainment
 * @property {number} shopping - Shopping
 * @property {number} subscriptions - Subscriptions
 * @property {number} other - Other wants
 */

/**
 * @typedef {Object} SavingsCategory
 * @property {number} emergency - Emergency fund
 * @property {number} sip - SIP/Investments
 * @property {number} fd_rd - FD/RD/Fixed Income
 * @property {number} goals - Goal-based savings
 */

/**
 * @typedef {Object} Categories
 * @property {NeedsCategory} needs - Needs breakdown
 * @property {WantsCategory} wants - Wants breakdown
 * @property {SavingsCategory} savings - Savings breakdown
 */

/**
 * @typedef {Object} Alert
 * @property {string} code - Alert code
 * @property {string} message - Alert message
 * @property {string} severity - Severity level
 * @property {string} suggestion - Suggestion
 */

/**
 * @typedef {Object} Metadata
 * @property {string} [city] - City name
 * @property {string} city_tier - Tier
 * @property {number} col_multiplier - COL multiplier
 * @property {string} [notes] - Additional notes
 */

/**
 * @typedef {Object} BudgetPlan
 * @property {number} income - Monthly income
 * @property {BudgetSplit} budget_split - Percentage split
 * @property {BudgetAmounts} budget_amounts - Amount split
 * @property {Categories} categories - Category breakdown
 * @property {Array<Alert>} alerts - Risk alerts
 * @property {string} explanation - AI-generated explanation
 * @property {Metadata} metadata - Metadata
 */

/**
 * @typedef {Object} BudgetGenerateResponse
 * @property {boolean} success - Success flag
 * @property {BudgetPlan} plan - Generated budget plan
 * @property {string} [error] - Error message if failed
 */

/**
 * @typedef {Object} BudgetRebalanceRequest
 * @property {Object} edited_plan - Edited plan amounts
 * @property {Object} original_inputs - Original user inputs
 * @property {string} city_tier - Tier
 * @property {number} col_multiplier - COL multiplier
 */

/**
 * @typedef {Object} BudgetRebalanceResponse
 * @property {boolean} success - Success flag
 * @property {BudgetPlan} plan - Rebalanced plan
 * @property {Array<Alert>} alerts - Updated alerts
 * @property {string} reasoning - Rebalance reasoning
 * @property {Metadata} metadata - Metadata
 */

// ============================================
// LOCALSTORAGE SCHEMAS
// ============================================

/**
 * @typedef {Object} SavedBudget
 * @property {Object} inputs - All user inputs
 * @property {string} mode - Budget mode used
 * @property {BudgetPlan} plan - Generated plan
 * @property {boolean} edited - Was plan edited
 * @property {string} timestamp - ISO timestamp
 * @property {Metadata} metadata - Metadata
 */

/**
 * @typedef {Object} BudgetHistory
 * Array of SavedBudget objects
 */

// ============================================
// VALIDATION SCHEMAS
// ============================================

export const validationRules = {
  monthly_income: {
    min: 10000, // Minimum income
    max: 10000000, // Maximum income
    message: 'Income must be between ₹10,000 and ₹1,00,00,000',
  },
  family_size: {
    min: 1,
    max: 10,
    message: 'Family size must be between 1 and 10',
  },
  col_multiplier: {
    min: 0.5,
    max: 2.0,
    message: 'COL multiplier must be between 0.5 and 2.0',
  },
  tenure_months: {
    min: 1,
    max: 360,
    message: 'Loan tenure must be between 1 and 360 months',
  },
  interest_rate: {
    min: 0,
    max: 50,
    message: 'Interest rate must be between 0% and 50%',
  },
  priority: {
    min: 1,
    max: 5,
    message: 'Priority must be between 1 and 5',
  },
};

// ============================================
// DEFAULT VALUES
// ============================================

export const defaults = {
  mode: 'smart_balanced',
  currency: 'INR',
  family_size: 1,
  lifestyle: 'moderate',
  city_tier: 'tier_1',
  col_multiplier: 1.0,
  budget_split: {
    needs_percent: 50,
    wants_percent: 30,
    savings_percent: 20,
  },
  minimum_savings_percent: 5,
  goals_limit: 5,
  loans_limit: 5,
  history_limit: 10,
};

// ============================================
// ALERT CODES
// ============================================

export const alertCodes = {
  HIGH_RENT_RATIO: 'HIGH_RENT_RATIO',
  HIGH_EMI_BURDEN: 'HIGH_EMI_BURDEN',
  NEGATIVE_CASHFLOW: 'NEGATIVE_CASHFLOW',
  LOW_SAVINGS_RATE: 'LOW_SAVINGS_RATE',
  HIGH_WANTS_SPENDING: 'HIGH_WANTS_SPENDING',
  INSUFFICIENT_EMERGENCY: 'INSUFFICIENT_EMERGENCY',
  VERY_LOW_INCOME: 'VERY_LOW_INCOME',
  INVALID_CITY: 'INVALID_CITY',
};

// ============================================
// SEVERITY LEVELS
// ============================================

export const severityLevels = {
  INFO: 'info',
  WARNING: 'warning',
  MODERATE: 'moderate',
  HIGH: 'high',
  CRITICAL: 'critical',
};

// ============================================
// BUDGET MODES
// ============================================

export const budgetModes = {
  BASIC: 'basic',
  AGGRESSIVE_SAVINGS: 'aggressive_savings',
  SMART_BALANCED: 'smart_balanced',
};

// ============================================
// LOCALSTORAGE KEYS
// ============================================

export const storageKeys = {
  lastPlan: 'vegakash.budget.lastPlan',
  history: 'vegakash.budget.history',
};

// ============================================
// EXPORT
// ============================================

export default {
  validationRules,
  defaults,
  alertCodes,
  severityLevels,
  budgetModes,
  storageKeys,
};

File: backend/budget_schemas/budget_planner.py
"""
Budget Planner V1.2 - Pydantic Schemas
Models for request/response validation and serialization
"""

from pydantic import BaseModel, Field, validator
from typing import List, Optional, Dict, Any
from datetime import datetime


# ============================================
# INPUT SCHEMAS
# ============================================

class IncomeInput(BaseModel):
    """Income and currency input"""
    monthly_income: float = Field(..., gt=10000, le=10000000, description="Monthly income in currency")
    currency: str = Field(default="INR", description="Currency code (INR, USD, EUR, GBP, etc.)")


class CityInput(BaseModel):
    """City and cost-of-living input"""
    country: Optional[str] = Field(None, description="Country name")
    state: Optional[str] = Field(None, description="State/Province name")
    city: Optional[str] = Field(None, description="City name")
    city_tier: str = Field(default="tier_1", description="City tier (tier_1, tier_2, tier_3, other)")
    col_multiplier: float = Field(default=1.0, ge=0.5, le=2.0, description="Cost-of-living multiplier")


class HouseholdInput(BaseModel):
    """Household and lifestyle input"""
    family_size: int = Field(default=1, ge=1, le=10, description="Number of family members")
    lifestyle: str = Field(default="moderate", description="Lifestyle type (minimal, moderate, comfort, premium)")


class FixedExpenses(BaseModel):
    """Fixed expense categories"""
    rent: float = Field(default=0, ge=0, description="Monthly rent/mortgage")
    utilities: float = Field(default=0, ge=0, description="Utilities (water, electricity, internet)")
    insurance: float = Field(default=0, ge=0, description="Insurance premiums")
    medical: float = Field(default=0, ge=0, description="Medical expenses")
    other: float = Field(default=0, ge=0, description="Other fixed expenses")


class VariableExpenses(BaseModel):
    """Variable expense categories"""
    groceries: float = Field(default=0, ge=0, description="Grocery expenses")
    transport: float = Field(default=0, ge=0, description="Transport/fuel")
    subscriptions: float = Field(default=0, ge=0, description="Subscriptions")
    entertainment: float = Field(default=0, ge=0, description="Entertainment")
    shopping: float = Field(default=0, ge=0, description="Shopping")
    dining_out: float = Field(default=0, ge=0, description="Dining out")
    other: float = Field(default=0, ge=0, description="Other variable expenses")


class LoanInput(BaseModel):
    """Loan/EMI input"""
    principal: float = Field(..., gt=0, description="Original loan amount")
    rate: float = Field(..., ge=0, le=50, description="Annual interest rate (%)")
    tenure_months: int = Field(..., ge=1, le=360, description="Remaining tenure in months")
    issuer: Optional[str] = Field(None, description="Bank/NBFC name")


class SavingsGoal(BaseModel):
    """Savings goal input"""
    name: str = Field(..., min_length=1, description="Goal name")
    target: float = Field(..., gt=0, description="Target amount")
    target_months: int = Field(..., ge=1, le=360, description="Timeline in months")
    priority: int = Field(default=3, ge=1, le=5, description="Priority (1-5)")


class BudgetGenerateRequest(BaseModel):
    """Complete budget generation request"""
    monthly_income: float = Field(..., gt=10000, le=10000000)
    currency: str = Field(default="INR")
    country: Optional[str] = None
    state: Optional[str] = None
    city: Optional[str] = None
    city_tier: str = Field(default="tier_1")
    col_multiplier: float = Field(default=1.0, ge=0.5, le=2.0)
    family_size: int = Field(default=1, ge=1, le=10)
    lifestyle: str = Field(default="moderate")
    fixed_expenses: FixedExpenses = Field(default_factory=FixedExpenses)
    variable_expenses: VariableExpenses = Field(default_factory=VariableExpenses)
    loans: List[LoanInput] = Field(default_factory=list, max_items=5)
    goals: List[SavingsGoal] = Field(default_factory=list, max_items=5)
    mode: str = Field(default="smart_balanced", description="Budget mode (basic, aggressive_savings, smart_balanced)")


# ============================================
# OUTPUT SCHEMAS
# ============================================

class BudgetSplit(BaseModel):
    """Budget percentage split"""
    needs_percent: float = Field(..., ge=0, le=100)
    wants_percent: float = Field(..., ge=0, le=100)
    savings_percent: float = Field(..., ge=0, le=100)


class BudgetAmounts(BaseModel):
    """Budget absolute amounts"""
    needs: float = Field(..., ge=0)
    wants: float = Field(..., ge=0)
    savings: float = Field(..., ge=0)


class NeedsCategory(BaseModel):
    """Needs subcategory breakdown"""
    rent: float = Field(default=0, ge=0)
    utilities: float = Field(default=0, ge=0)
    groceries: float = Field(default=0, ge=0)
    transport: float = Field(default=0, ge=0)
    insurance: float = Field(default=0, ge=0)
    medical: float = Field(default=0, ge=0)
    emi: float = Field(default=0, ge=0)
    other: float = Field(default=0, ge=-10000)  # Allow negative for rebalancing


class WantsCategory(BaseModel):
    """Wants subcategory breakdown"""
    dining: float = Field(default=0, ge=0)
    entertainment: float = Field(default=0, ge=0)
    shopping: float = Field(default=0, ge=0)
    subscriptions: float = Field(default=0, ge=0)
    other: float = Field(default=0, ge=0)


class SavingsCategory(BaseModel):
    """Savings subcategory breakdown"""
    emergency: float = Field(default=0, ge=0)
    sip: float = Field(default=0, ge=0)
    fd_rd: float = Field(default=0, ge=0)
    goals: float = Field(default=0, ge=0)


class Categories(BaseModel):
    """Complete category breakdown"""
    needs: NeedsCategory
    wants: WantsCategory
    savings: SavingsCategory


class Alert(BaseModel):
    """Alert/Risk detection"""
    code: str = Field(..., description="Alert code")
    message: str = Field(..., description="Alert message")
    severity: str = Field(..., description="Severity level (info, warning, moderate, high, critical)")
    suggestion: str = Field(..., description="Actionable suggestion")


class Metadata(BaseModel):
    """Budget plan metadata"""
    city: Optional[str] = None
    city_tier: str = Field(default="tier_1")
    col_multiplier: float = Field(default=1.0)
    notes: Optional[str] = None


class BudgetPlan(BaseModel):
    """Complete budget plan output"""
    income: float = Field(..., gt=0)
    budget_split: BudgetSplit
    budget_amounts: BudgetAmounts
    categories: Categories
    alerts: List[Alert] = Field(default_factory=list)
    explanation: str = Field(..., description="AI-generated explanation")
    metadata: Metadata


class BudgetGenerateResponse(BaseModel):
    """Budget generation API response"""
    success: bool = True
    plan: Optional[BudgetPlan] = None
    error: Optional[str] = None


class BudgetRebalanceRequest(BaseModel):
    """Rebalance budget request"""
    edited_plan: Dict[str, Any] = Field(...)
    original_inputs: BudgetGenerateRequest = Field(...)
    city_tier: str = Field(default="tier_1")
    col_multiplier: float = Field(default=1.0)


class BudgetRebalanceResponse(BaseModel):
    """Rebalance budget API response"""
    success: bool = True
    plan: Optional[BudgetPlan] = None
    alerts: List[Alert] = Field(default_factory=list)
    reasoning: Optional[str] = None
    metadata: Optional[Metadata] = None
    error: Optional[str] = None


class SavedBudget(BaseModel):
    """Saved budget in LocalStorage"""
    inputs: BudgetGenerateRequest
    mode: str
    plan: BudgetPlan
    edited: bool = False
    timestamp: datetime = Field(default_factory=datetime.utcnow)
    metadata: Metadata


# ============================================
# VALIDATORS
# ============================================

class BudgetGenerateRequest(BudgetGenerateRequest):
    """Extended with validators"""
    
    @validator('lifestyle')
    def validate_lifestyle(cls, v):
        valid_lifestyles = ['minimal', 'moderate', 'comfort', 'premium']
        if v not in valid_lifestyles:
            raise ValueError(f"Lifestyle must be one of {valid_lifestyles}")
        return v
    
    @validator('mode')
    def validate_mode(cls, v):
        valid_modes = ['basic', 'aggressive_savings', 'smart_balanced']
        if v not in valid_modes:
            raise ValueError(f"Mode must be one of {valid_modes}")
        return v
    
    @validator('city_tier')
    def validate_tier(cls, v):
        valid_tiers = ['tier_1', 'tier_2', 'tier_3', 'other']
        if v not in valid_tiers:
            raise ValueError(f"City tier must be one of {valid_tiers}")
        return v
